### PR TITLE
Upgrade into Rails 7.0 and Ruby 3.1

### PR DESCRIPTION
**Pull request by bivanalhar**

### CODE REVIEW

File: spec/support/acts_as_tenant.rb [hunk 1]
```diff
@@ -19,9 +19,15 @@ def with_tenant(tenant, &proc)
       context "with tenant #{tenant.inspect}" do |*params|
         before(:each) do
           ActsAsTenant.current_tenant = send(tenant)
+
+          # it's recommended to use test_tenant as it's thread-safe for
+          # testing in parallel settings, which we do.
+          # Please refer to https://github.com/ErwinM/acts_as_tenant/issues/277
+          ActsAsTenant.test_tenant = send(tenant)
```
Comment (cysjonathan) [inline spec/support/acts_as_tenant.rb:26]: `send(tenant)` called twice, isn't it better to store it as a variable then use that variable twice instead?

File: app/models/course/assessment.rb [hunk 2]
```diff
@@ -103,11 +103,8 @@ class Course::Assessment < ApplicationRecord
   # @!method self.ordered_by_date_and_title
   #   Orders the assessments by the starting date and title.
   scope :ordered_by_date_and_title, (lambda do
-    select(<<~SQL).
-      course_assessments.*, course_reference_times.start_at, course_lesson_plan_items.title
-    SQL
     joins(:lesson_plan_item).
-    merge(Course::LessonPlan::Item.ordered_by_date_and_title)
```
Comment (cysjonathan) [inline app/models/course/assessment.rb:110]: Why are we removing some SQL here?

File: app/models/course/forum/topic.rb [hunk 3]
```diff
@@ -44,10 +44,8 @@ class Course::Forum::Topic < ApplicationRecord
   # @!attribute [r] post_count
   #   The number of posts in this topic.
   calculated :post_count, (lambda do
-    Course::Discussion::Topic.joins(:posts).
-      where('actable_id = course_forum_topics.id').
-      where(actable_type: Course::Forum::Topic.name).
-      select("count('*')")
```
Comment (cysjonathan) [inline app/models/course/forum/topic.rb:50]: Try not to commit style changes for upgrades like these, it'll end up causing more clutter when reviewing.

File: app/models/course/forum/topic.rb [hunk 4]
```diff
@@ -79,11 +77,14 @@ class Course::Forum::Topic < ApplicationRecord
           group('course_discussion_posts.topic_id').
           where(topic_id: topic_ids)
 
-    last_posts = Course::Discussion::Post.with_creator.where('id in (?) or id in (?)', min_ids, max_ids)
+    last_posts = Course::Discussion::Post.with_creator.
+          where('course_discussion_posts.id in (?) or course_discussion_posts.id in (?)', min_ids, max_ids)
```
Comment (cysjonathan) [inline app/models/course/forum/topic.rb:81]: Why the change in expression? Does the previous one not work in Rails 7?

File: app/models/course/lesson_plan/item.rb [hunk 5]
```diff
@@ -65,14 +65,16 @@ class Course::LessonPlan::Item < ApplicationRecord
   scope :with_personal_times_for, (lambda do |course_user|
     personal_times =
       if course_user.nil?
-        []
+        nil
       else
-        Course::PersonalTime.where(course_user_id: course_user.id, lesson_plan_item_id: all).to_a
+        Course::PersonalTime.where(course_user_id: course_user.id, lesson_plan_item_id: all)
```
Comment (cysjonathan) [inline app/models/course/lesson_plan/item.rb:70]: Why change the function signature here?

File: app/models/course/video.rb [hunk 6]
```diff
@@ -51,8 +51,7 @@ class Course::Video < ApplicationRecord
   # @!method self.ordered_by_date_and_title
   #   Orders the videos by the starting date and title.
   scope :ordered_by_date_and_title, (lambda do
-    select('course_videos.*, course_reference_times.start_at, course_lesson_plan_items.title').
-      joins(:lesson_plan_item).
+    joins(:lesson_plan_item).
```
Comment (cysjonathan) [inline app/models/course/video.rb:54]: Why are we removing the `select`?

File: bin/setup [hunk 7]
```diff
@@ -9,8 +9,8 @@ def system!(*args)
 end
 
 FileUtils.chdir APP_ROOT do
-  # This script is a way to setup or update your development environment automatically.
-  # This script is idempotent, so that you can run it at anytime and get an expectable outcome.
+  # This script is a way to set up or update your development environment automatically.
+  # This script is idempotent, so that you can run it at any time and get an expectable outcome.
```
Comment (cysjonathan) [inline bin/setup:13]: this isn't important to include in rails 7 upgrade, let's avoid makin irrelevant changes

File: config.ru [hunk 8]
```diff
@@ -3,3 +3,4 @@
 
 require ::File.expand_path('config/environment', __dir__)
 run Rails.application
+Rails.application.load_server
```
Comment (cysjonathan) [inline config.ru:6]: Why is this suddenly required?

File: config/application.rb [hunk 9]
```diff
@@ -8,14 +8,12 @@
 Bundler.require(*Rails.groups)
 
 # Load dotenv only in development environment
-Dotenv::Railtie.load if ['development', 'test'].include? ENV['RAILS_ENV']
+Dotenv::Rails.load if ['development'].include? ENV['RAILS_ENV']
```
Comment (cysjonathan) [inline config/application.rb:11]: This file has many changes but not enough documentation of why it is changed.

File: config/environments/production.rb [hunk 10]
```diff
@@ -49,9 +61,9 @@
   # Force all access to the app over SSL, use Strict-Transport-Security, and use secure cookies.
   # config.force_ssl = true
 
-  # Use the lowest log level to ensure availability of diagnostic information
-  # when problems arise.
-  config.log_level = :debug
+  # Include generic and useful information about system operation, but avoid logging too much
+  # information to avoid inadvertent exposure of personally identifiable information (PII).
+  config.log_level = :info
```
Comment (cysjonathan) [inline config/environments/production.rb:66]: Why is this being raised from debug to info?

File: config/environments/test.rb [hunk 11]
```diff
@@ -13,15 +15,18 @@
   # just for the purpose of running a single test. If you are using a tool that
   # preloads Rails for running tests, you may have to set it to true.
   # NOTE: set it to false would result in CI tests to timeout
-  config.eager_load = true
+  config.eager_load = false
```
Comment (cysjonathan) [inline config/environments/test.rb:18]: Explain the changes in this file?

File: spec/services/course/duplication/course_duplication_service_spec.rb [hunk 12]
```diff
@@ -410,12 +410,7 @@
 
         it 'duplicates achievement with attached badge' do
           new_achievement_with_badge = new_course.achievements.select(&:badge_url).first
-          new_achievement_id = new_achievement_with_badge.id
-          badge_folder = Rails.root.join('public', 'uploads', 'images', 'course',
-                                         'achievement', new_achievement_id.to_s, 'badge')
-          badge_file = badge_folder.join('minion.png')
-          expect(File.directory?(badge_folder)).to be true
-          expect(File.exist?(badge_file)).to be true
+          expect(File.exist?(File.join(Rails.root, 'public', new_achievement_with_badge.badge.url))).to be true
```
Comment (cysjonathan) [inline spec/services/course/duplication/course_duplication_service_spec.rb:413]: better to split this into
```rb
badge_file = ...
expect(File.exist?(badge_file)).to be true
```

1. fewer code changes
2. explicitly declare badge file path

File: config/initializers/backtrace_silencers.rb [hunk 13]
```diff
@@ -2,7 +2,8 @@
...
 
 # You can add backtrace silencers for libraries that you're using but don't wish to see in your backtraces.
-# Rails.backtrace_cleaner.add_silencer { |line| line =~ /my_noisy_library/ }
+# Rails.backtrace_cleaner.add_silencer { |line| /my_noisy_library/.match?(line) }
 
-# You can also remove all the silencers if you're trying to debug a problem that might stem from framework code.
-# Rails.backtrace_cleaner.remove_silencers!
+# You can also remove all the silencers if you're trying to debug a problem that might stem from framework code
+# by setting BACKTRACE=1 before calling your invocation, like "BACKTRACE=1 ./bin/rails runner 'MyClass.perform'".
+Rails.backtrace_cleaner.remove_silencers! if ENV['BACKTRACE']
```
Comment (cysjonathan) [inline config/initializers/backtrace_silencers.rb:9]: Do we use `BACKTRACE` the env variable? Is this required for rails 7 upgrade?

File: config/initializers/new_framework_defaults_6_0.rb [hunk 14]
```diff
@@ -1,46 +0,0 @@
...
-# will be removed in Rails 6.1. This setting is not backwards compatible with earlier Rails versions.
-# If you send mail in the background, job workers need to have a copy of
-# MailDeliveryJob to ensure all delivery jobs are processed properly.
-# Make sure your entire app is migrated and stable on 6.0 before using this setting.
-Rails.application.config.action_mailer.delivery_job = 'ActionMailer::MailDeliveryJob'
-
-# Enable the same cache key to be reused when the object being cached of type
-# `ActiveRecord::Relation` changes by moving the volatile information (max updated at and count)
-# of the relation's cache key into the cache version to support recycling cache key.
-Rails.application.config.active_record.collection_cache_versioning = true
```
Comment (cysjonathan) [inline config/initializers/new_framework_defaults_6_0.rb:46]: Deleting config/initializers/new_framework_defaults_6_0.rb and 5_2.rb, is there no need to create a new defaults for 7_0?

File: config/initializers/permissions_policy.rb [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+# information see https://developers.google.com/web/updates/2018/06/feature-policy
+#
+# Rails.application.config.permissions_policy do |f|
+#   f.camera      :none
+#   f.gyroscope   :none
+#   f.microphone  :none
+#   f.usb         :none
+#   f.fullscreen  :self
+#   f.payment     :self, "https://secure.example.com"
+# end
```
Comment (cysjonathan) [inline config/initializers/permissions_policy.rb:12]: Do we need a permission policy? If yes, then shouldn't we uncomment these, otherwise, why include the file?

File: Gemfile [hunk 16]
```diff
@@ -39,14 +43,18 @@ gem 'active_record_upsert', '0.11.2'
 gem 'friendly_id'
 
 # HTML Pipeline and dependencies
-gem 'html-pipeline'
+gem 'html-pipeline', '2.14.3'
```
Comment (cysjonathan) [inline Gemfile:46]: Remove the version locks. Imposing a version lock means we will no longer be notified of future upstream updates.  
You can upgrade the gems one-by-one to avoid unnecessary upgrades.

File: app/controllers/concerns/application_multitenancy.rb [hunk 17]
```diff
@@ -3,7 +3,9 @@ module ApplicationMultitenancy
   private
 
   def deduce_and_set_current_tenant
-    set_current_tenant(deduce_tenant)
+    tenant = deduce_tenant
+    ActsAsTenant.current_tenant = tenant
+    ActsAsTenant.test_tenant = tenant
```
Comment (cysjonathan) [inline app/controllers/concerns/application_multitenancy.rb:8]: This doesn't look right. Why would we need `.test_tenant` in production?

**Review by cysjonathan — CHANGES_REQUESTED** [review]

File: spec/support/acts_as_tenant.rb (hunk 1 above)
Comment (bivanalhar) [reply spec/support/acts_as_tenant.rb:26]: done

File: app/models/course/assessment.rb (hunk 2 above)
Comment (bivanalhar) [reply app/models/course/assessment.rb:110]: this SQL is redundant since actually in the statement that follows (`merge(Course::LessonPlan::Item.ordered_by_date_and_title)`), we already covers `includes(reference_times: :reference_timeline)` and the `lesson_plan_items` itself is already covered inside `joins` statement

File: app/models/course/video.rb (hunk 6 above)
Comment (bivanalhar) [reply app/models/course/video.rb:54]: same as `models/course/assessment.rb`, this select is quite redundant

File: app/models/course/forum/topic.rb (hunk 4 above)
Comment (bivanalhar) [reply app/models/course/forum/topic.rb:81]: this is weird. I was working on this while doing the change in Rails 7, and the error was thrown here mentioning that `id is ambiguous` (between the `id` of `course_discussion_posts` and the `id` of `course_forum_topics`), but seems like after I try again, this error is gone already.

Reverting back to the original one.

File: app/models/course/lesson_plan/item.rb (hunk 5 above)
Comment (bivanalhar) [reply app/models/course/lesson_plan/item.rb:70]: it's because the previous one that we are using (`ManualPreloader`) is no longer compatible with the `Preloader` object that's used inside `Active Record` in Rails 7, and after checking the object itself, the execution call of `ManualPreloader` only calls `super`, means that it's actually doing the same thing as `ActiveRecord::Association::Preloader`.

Hence, I decided to just use `ActiveRecord::Association::Preloader` here instead, and with it comes the change in the argument signature that I need to use. One of them being needing to provide the `scope` variable inside the `Preloader`, in which after I reviewed about how the `scope` should be defined in our codebase, usually does not have `to_a` and that's why I changed it

(also, the default value of `scope` should it not be provided is merely `nil`, not `[]`)

File: bin/setup (hunk 7 above)
Comment (bivanalhar) [reply bin/setup:13]: I think it should be fine? Since it's merely English word fixup, and if we let it be like the previously it will end up not being attended anymore

File: config.ru (hunk 8 above)
Comment (bivanalhar) [reply config.ru:6]: I checked again and in Rails 7, it's actually okay to not including this (I remembered in Rails 6.1 the application does not run properly if this line is not included)

Reverting back

File: config/application.rb (hunk 9 above)
Comment (bivanalhar) [reply config/application.rb:11]: for this specific line, it's because of the Dotenv library upgrade, and when we're upgrading to `3.1.0` last time, this `Dotenv::Railtie` should have been changed to `Dotenv::Rails` since the former one is marked as deprecated already. Refer to https://github.com/bkeepers/dotenv/blob/v3.1.2/lib/dotenv/rails.rb#L113

File: config/environments/test.rb (hunk 11 above)
Comment (bivanalhar) [reply config/environments/test.rb:18]: this is actually because the comment just up there explains that the `eager_load` should be set as `false`. And now that we have moved to `Zeitwerk`, which strongly promotes the efficiency in doing the loading, I tried to change it back to what it should be (`false`) and now nothing breaks so I decided to change it like that

Though, I should have deleted that NOTE so as not to confuse future dev

File: spec/services/course/duplication/course_duplication_service_spec.rb (hunk 12 above)
Comment (bivanalhar) [reply spec/services/course/duplication/course_duplication_service_spec.rb:413]: done

File: config/initializers/permissions_policy.rb (hunk 15 above)
Comment (bivanalhar) [reply config/initializers/permissions_policy.rb:12]: Sorry for the oversight. It was changed before I took over this Rails Upgrade, and I was not aware of this file.. Checking on the link provided over there, and also on Rails 7 Documentation, this permission_policy is merely recommended to up our security level but not required

And also for our use case, I think we don't need it for now, so we're safe to delete this file

File: config/initializers/new_framework_defaults_6_0.rb (hunk 14 above)
Comment (bivanalhar) [reply config/initializers/new_framework_defaults_6_0.rb:46]: we don't need that, since our app only supports one version at a time
This new framework file is only relevant if we build the library, in which we need to support multiple version (for example, developing our own self-maintained gem in which to ensure smoothness in upgrading the Rails system, we need that gem to support both Rails 6 and Rails 7)

File: Gemfile (hunk 16 above)
Comment (bivanalhar) [reply Gemfile:46]: Done

File: app/controllers/concerns/application_multitenancy.rb (hunk 17 above)
Comment (bivanalhar) [reply app/controllers/concerns/application_multitenancy.rb:8]: oh sorry, actually I made this when I fix the bug in Rspec, and yes this `test_tenant` is only relevant when used in our unit test, even though it won't affect production much

(the only relevance for `test_tenant` is only if we want to get the current_tenant, which is always set as `Current.current_tenant || test_tenant || || default_tenant`)

nevertheless, I'll set this one so that it'll change only if the RAILS_ENV is production

File: app/models/course/forum/topic.rb (hunk 3 above)
Comment (bivanalhar) [reply app/models/course/forum/topic.rb]: Done

File: config/environments/production.rb (hunk 10 above)
Comment (bivanalhar) [reply config/environments/production.rb:66]: the comment right before this line explains the reasoning behind the raising of the level
And also, in Rails 7.1 (which will be done after this upgrade), the default `log_level` will be set to `:info` in production, and hence it should be good practice to start it from now.

https://guides.rubyonrails.org/v7.0.8.4/debugging_rails_applications.html

File: config/initializers/backtrace_silencers.rb (hunk 13 above)
Comment (bivanalhar) [reply config/initializers/backtrace_silencers.rb:9]: inspired by the following
https://github.com/rails/rails/pull/50563

this has been merged, but it will be made effective starting Rails 7.2..
This environment variable is useful if we want to debug the framework code easier

File: config/environments/production.rb (hunk 10 above)
Comment (cysjonathan) [reply config/environments/production.rb:66]: Ok, but will need to monitor the impact on the amount of logs we keep.  
Previously we had an issue with lograge generating a lot of logs which cased our AWS bill to spike.

File: Gemfile.lock [hunk 18]
```diff
@@ -134,9 +142,7 @@ GEM
       activesupport (>= 3.0.0)
       uniform_notifier (~> 1.11)
     byebug (11.1.3)
-    calculated_attributes (0.5.0)
-      activerecord (>= 3.2.20)
-    cancancan (3.6.1)
+    cancancan (3.5.0)
```
Comment (cysjonathan) [inline Gemfile.lock]: `cancancan` downgraded?

File: app/models/course/assessment.rb (hunk 2 above)
Comment (cysjonathan) [reply app/models/course/assessment.rb:110]: Does this affect the amount of information we fetch? e.g. `course_lesson_plan_items.title` vs `course_lesson_plan_items.*`?

File: app/models/course/video.rb (hunk 6 above)
Comment (cysjonathan) [reply app/models/course/video.rb:54]: Does this affect the amount of information we fetch?

File: config/application.rb (hunk 9 above)
Comment (cysjonathan) [reply config/application.rb:11]: Alright, and why the removal of `'test'` here?

File: Gemfile [hunk 19]
```diff
@@ -195,7 +197,7 @@ gem 'rubyzip', require: 'zip'
 gem 'nokogiri', '>= 1.8.1'
 
 # Polyglot support
-gem 'coursemology-polyglot'
+gem 'coursemology-polyglot', '0.3.8'
```
Comment (cysjonathan) [inline Gemfile:203]: why lock to `0.3.8`?

File: Gemfile [hunk 20]
```diff
@@ -50,8 +50,12 @@ gem 'rinku'
 gem 'rouge', '~> 3'
 gem 'ruby-oembed'
 
+# to help obtaining app wide URI that uniquely identifies model instance
+# (used in notify_identifier for NOTIFY/LISTEN to jobs)
+gem 'globalid'
+
 # Build JSON APIs with ease. Read more: https://github.com/rails/jbuilder
-gem 'jbuilder'
+gem 'jbuilder', '2.12.0'
```
Comment (cysjonathan) [inline Gemfile:57]: why lock this version?

File: Gemfile [hunk 21]
```diff
@@ -100,7 +104,7 @@ end
 
 group :development, :test do
   # Use RSpec for Behaviour testing
-  gem 'rspec-rails'
+  gem 'rspec-rails', '6.1.2'
```
Comment (cysjonathan) [inline Gemfile:106]: why lock this version?

File: config/environments/test.rb [hunk 22]
```diff
@@ -50,9 +49,6 @@
   # We will assume that we are running on localhost
   config.action_mailer.default_url_options = { host: 'localhost:3200' }
 
-  # Use the threaded background job adapter for replicating the production environment.
-  config.active_job.queue_adapter = :background_thread
-
```
Comment (cysjonathan) [inline config/environments/test.rb]: What about this part of the file that's changed? we don't need threaded background job anymore?

File: spec/models/course/assessment_spec.rb [hunk 23]
```diff
@@ -137,7 +137,7 @@
           subject.save
 
           expect(subject.folder.name).to eq(new_title)
-          expect(subject.folder.start_at).to eq(new_start_at)
+          expect(subject.folder.start_at).to be_within(0.000001).of(new_start_at)
```
Comment (cysjonathan) [inline spec/models/course/assessment_spec.rb:140]: https://github.com/Coursemology/coursemology2/blob/2571b6c1d960e02afa382c0f2cbcd1a61c63a34c/spec/jobs/course/assessment/closing_reminder_job_spec.rb#L17

Saw one other spec use `.to_i` to drop the sub-second precision.

https://github.com/search?q=repo%3ACoursemology%2Fcoursemology2%20be_within&type=code

Others seem to prefer using `be_within(1.second)`, better to keep code aligned.

File: app/models/course/lesson_plan/item.rb (hunk 5 above)
Comment (cysjonathan) [reply app/models/course/lesson_plan/item.rb:70]: https://github.com/Coursemology/coursemology2/pull/7427/commits/e88c34b157737fd011b290b509fa1ae1475a6ca5
^ commit should be squashed with current commit.

File: spec/support/acts_as_tenant.rb [hunk 24]
```diff
@@ -18,10 +18,17 @@ module ModelHelpers
     def with_tenant(tenant, &proc)
       context "with tenant #{tenant.inspect}" do |*params|
         before(:each) do
-          ActsAsTenant.current_tenant = send(tenant)
+          cur_tenant = send(tenant)
```
Comment (cysjonathan) [inline spec/support/acts_as_tenant.rb:21]: use `current_tenant` instead of `cur_tenant`, follow the variable naming convention.

File: app/controllers/concerns/application_multitenancy.rb (hunk 17 above)
Comment (cysjonathan) [reply app/controllers/concerns/application_multitenancy.rb]: Looked at the proposed newer code, `if ENV['RAILS_ENV'] == 'test'` I think we can just omit that, since assignment isn't costly.  We can just assign and reduce the use of magic string `'test'` in the code.

https://github.com/Coursemology/coursemology2/blob/2571b6c1d960e02afa382c0f2cbcd1a61c63a34c/config/initializers/action_cable_acts_as_tenant.rb#L8

Also isn't this now dead code? Need to clean up.

File: Gemfile (hunk 21 above)
Comment (bivanalhar) [reply Gemfile:106]: According to its [gem repo](https://github.com/rspec/rspec-rails), specific Rails version requires specific Rspec version as well, and according to their policy, version 6.1 and 7.0 should be kept at Rspec 6.1 (though, it's not sure about their plan for Rails 7.1, whether it's going to be RSpec 6.x or Rspec 7.x), that's why I locked it to 6.1.

The newest version was just released less than a month ago (they have released it previously for 6.1.3 but they got bug and they just announced to rollback to 6.1.2. I tried to search for that announcements but I couldn't find it anymore)

Perhaps I should lock it to 6.x

File: Gemfile (hunk 19 above)
Comment (bivanalhar) [reply Gemfile:203]: This gem is we updating ourselves, and last time it's not locked, this gem got updated to 0.3.5 and created an error of cannot detect Python 3.12 object..

We need to update that polyglot gem first to denote 0.3.8 as the latest before we open the lock

File: Gemfile [hunk 25]
```diff
@@ -18,7 +18,7 @@ gem 'rack-cors'
 # Instance/Course settings
 gem 'settings_on_rails', git: 'https://github.com/bivanalhar/settings_on_rails'
 # Manage read/unread status
-gem 'unread'
+gem 'unread', '0.13.1'
```
[the remark this answers is not in the document]
Comment (bivanalhar) [reply Gemfile:21]: this one is purely oversight. I did this so that the `bundle install` command can fetch the latest version, but I did not revert it back once it's done. Will take note of this in the future not to leave it locked

File: Gemfile (hunk 20 above)
Comment (bivanalhar) [reply Gemfile:57]: this one is purely oversight. I did this so that the `bundle install` command can fetch the latest version, but I did not revert it back once it's done. Will take note of this in the future not to leave it locked

File: Gemfile (hunk 19 above)
Comment (cysjonathan) [reply Gemfile:203]: It shouldn't update to a lower version, should check what happened, don't lock so that it will auto bump to latest.

File: app/models/course/assessment.rb (hunk 2 above)
Comment (bivanalhar) [reply app/models/course/assessment.rb:110]: Not at all. This scope is merely to order the assessments based on date and title (firstly by date, then by title if the date is same). As long as the information regarding the `start_at` and `title` is provided, all is safe

The reason those are deleted was because inside `Course::LessonPlan::Item.ordered_by_date_and_title`, those informations extracted by the SQL (`course_reference_times.start_at` and `course_lesson_plan_items.title`) are also extracted there

File: app/models/course/video.rb (hunk 6 above)
Comment (bivanalhar) [reply app/models/course/video.rb:54]: not at all, also by the same reason as the one I've written for the same thing in `models/course/assessment.rb`

File: config/environments/production.rb (hunk 10 above)
Comment (bivanalhar) [reply config/environments/production.rb:66]: Roger that

File: Gemfile (hunk 19 above)
Comment (bivanalhar) [reply Gemfile:203]: I got the reason why this happened
I used the `bundle update` since this is what's suggested by Rails Dev for upgrading to the newer version, and now that I checked around, it's already an existing issue that running that command can lower the version, as mentioned here in https://github.com/rubygems/bundler/issues/1266.

It should be taken note for the other CM developers also, that if want to use this command, need to proceed with caution..

File: spec/models/course/assessment_spec.rb (hunk 23 above)
Comment (bivanalhar) [reply spec/models/course/assessment_spec.rb:140]: understand about this. Changing to `be_within(1.second)`
The reason I did it this way is because Rails 7 is the first version of Rails in which precision for datetime is 6 by default, and hence I want to keep it as precise as possible

But then it will create a hassle since for the sake of consistency, I also need to change all the datetime to be like that. You're right that it's better to keep the code aligned with the existing way to resolve the discrepancies

File: Gemfile (hunk 19 above)
Comment (cysjonathan) [reply Gemfile:203]: The issue mentions it's possible if there are conflicting dependencies specified, but `coursemology-polyglot` shouldn't be specific in any other gem as far as I can tell, so it's strange.  
Nevermind don't need to focus on this, just keep it unlocked and ensure it doesn't downgrade.

File: spec/support/acts_as_tenant.rb (hunk 24 above)
Comment (bivanalhar) [reply spec/support/acts_as_tenant.rb:21]: Done

File: Gemfile.lock (hunk 18 above)
Comment (bivanalhar) [reply Gemfile.lock]: sorry, it's because of the `bundle update` command that I used to automatically renew the version of our gems, but apparently some got downgraded even though I manually upgrade it later on

will squash all the commits for 6.1 into 1

File: app/models/course/lesson_plan/item.rb (hunk 5 above)
Comment (bivanalhar) [reply app/models/course/lesson_plan/item.rb:70]: Done

File: app/controllers/concerns/application_multitenancy.rb (hunk 17 above)
Comment (bivanalhar) [reply app/controllers/concerns/application_multitenancy.rb]: roger that, reverting to the original one
as for `ActionCable`, it's being referred by some other important module such as `ApplicationMultitenancy`, and since this very function is being referred by the `ClassMethod` defined inside that module, it's better to keep it there

File: config/environments/test.rb (hunk 22 above)
Comment (bivanalhar) [reply config/environments/test.rb]: oh sorry, when I was upgrading to Rails 7 this line creates the problem (due to the setting of `eager_load` inside test env to be set false), and hence at that time I tried removing this line and everything still works well

Now that I revisited this file, I realised that debugging way is wrong here; what should be done is to explicitly requiring that `BackgroundThreadAdapter` towards this file and still keep this config

I have done the reverting, and it will be available later after finishing all the changes needed

File: app/controllers/concerns/application_multitenancy.rb (hunk 17 above)
Comment (cysjonathan) [reply app/controllers/concerns/application_multitenancy.rb:8]: In that case, wouldn't it be better to update the function definition of `set_current_tenant` to set both `.current_tenant` and `.test_tenant`?

**Review by cysjonathan — APPROVED** [review]
